subscriptions-api-openapi.yaml

```diff
@@ -125,126 +132,349 @@ paths:
+              example: 1119920371
       required:
         - id
+        - sink
         - protocol
+    Protocol:
+      type: string
+      enum: ["HTTP", "MQTT3", "MQTT5", "AMQP", "NATS", "KAFKA"]
+      description: REQUIRED. Identifier of a delivery protocol.
+      example: "HTTP"
     Filter:
-      properties:
-        dialect:
-          type: string
-          description: Filter dialect
-          example: "simple"
-          default: "simple"
-        conditions:
-          type: array
-          description: Filter conditions
-          items:
-            type: object
+      title: "Filter"
+      description: "A filter from a selection of multiple filter types and dialects"
+      type: object
+      additionalProperties: false
+    AllFilter:
+      allOf:
+        - $ref: "#/components/schemas/Filter"
+        - type: object
+          description: "all filter" 
+          properties:
+            all:
+              description: "This filter evaluates to 'true' if all contained filters are 'true'"
+              type: array
+              minItems: 1
+              items:
+                title: "Filter entry"
+                oneOf:
+                - $ref: "#/components/schemas/Filter"
+                - $ref: "#/components/schemas/AllFilter"
+                - $ref: "#/components/schemas/AnyFilter"
+                - $ref: "#/components/schemas/NotFilter"
+                - $ref: "#/components/schemas/ExactFilter"
+                - $ref: "#/components/schemas/PrefixFilter"
+                - $ref: "#/components/schemas/SuffixFilter"
+                - $ref: "#/components/schemas/DialectFilter"
+          additionalProperties: false
+    AnyFilter:
+      allOf:
+        - $ref: "#/components/schemas/Filter"
+        - type: object
+          description: "any filter" 
+          properties:
+            any:
+              description: "This filter evaluates to 'true' if any contained filters are 'true'"
+              type: array
+              minItems: 1
+              items:
+                title: "Filter entry"
+                oneOf:
+                  - $ref: "#/components/schemas/Filter"
+                  - $ref: "#/components/schemas/AllFilter"
+                  - $ref: "#/components/schemas/AnyFilter"
+                  - $ref: "#/components/schemas/NotFilter"
+                  - $ref: "#/components/schemas/ExactFilter"
+                  - $ref: "#/components/schemas/PrefixFilter"
+                  - $ref: "#/components/schemas/SuffixFilter"
+                  - $ref: "#/components/schemas/DialectFilter"
+          additionalProperties: false
+    NotFilter:
+      allOf:
+        - $ref: "#/components/schemas/Filter"
+        - type: object
+          description: "not filter" 
+          properties:
+            not:
+              type: object
+              properties:
+                filter:
+                  title: "Filter entry"
+                  oneOf:
```

Comment:
More than repeating this `oneOf`, how about having the `oneOf` straight inside `#/components/schemas/Filter` and removing the `allOf` in the child schemes? It's the more "idiomatic" way of defining recursive data structures. Like:

```
Filter:
  title: "Filter"
  description: "A filter from a selection of multiple filter types and dialects"
  type: object
  oneOf:
    - $ref: "#/components/schemas/AllFilter"
    - $ref: "#/components/schemas/AnyFilter"
    - $ref: "#/components/schemas/NotFilter"
    - $ref: "#/components/schemas/ExactFilter"
    - $ref: "#/components/schemas/PrefixFilter"
    - $ref: "#/components/schemas/SuffixFilter"
    - $ref: "#/components/schemas/DialectFilter"
```

And:

```
    AnyFilter:
      type: object
      description: "any filter" 
      properties:
        any:
          description: "This filter evaluates to 'true' if any contained filters are 'true'"
          type: array
          minItems: 1
          items:
            - $ref: "#/components/schemas/Filter"
      additionalProperties: false
```